### ISSUE TITLE
Look at tsc200x-core power management

### ISSUE THREAD

**Issue by MerlijnWajer** [maemo-leste/bugtracker#321]
It currently relies on the `SIMPLE_DEV_PM_OPS` macro defined in `pm.h`, but I think it needs (a lot) more power management work. For one, the patch that we have that adds a disabled attribute has, I think, logic flaws: https://github.com/maemo-leste/n9xx-linux/commit/3345caef0c18872aac4329f5ab3870b954716d80#diff-1328c316f7a13705ecc6e1fbee8e9958R649

That looks wrong. If the ts is disabled, we should be able to suspend it. And if we enable it, we should wake it up again. I am wondering if we need to switch to `pm_runtime_` calls instead, as done in (for example) `cyttsp4-core.c`.

This should allow us to not always have ts on, save us the power of always having it on, and also hopefully bring us closer to allowing the SoC to enter deeper idle modes.

**Comment by MerlijnWajer** [comment]
We might want to take a similar look at `atmel_mxt_ts.c`

**Comment by spinal84** [comment]
I did some testing on Fremantle vs Leste.
Enabling/disabling touchscreen while tk is locked doesn't change power consumption much.
In both cases it was about 30-70mA with wifi on.
While Leste is 150-180mA in the similar conditions.

**Comment by MerlijnWajer** [comment]
Right, but if it prevents off-mode, that is a problem. Do you have wifi power saving on?

All of the components have to play nice for the N900 and Droid4 to enter off mode, hence the ticket.

**Comment by spinal84** [comment]
What I did I tried to use a logic to understand how countless lines of code work together. And how the "black box" which is Nokia N900 device works.

We can't put out time and attention everywhere in hope it helps somehow to fix things. It's mandatory to find and fix the real roots of issues. What you are talking about is an assumption that needs to be checked first before we go deeper with it. We need to check somehow that your assumption is correct. Right?

We can't rely on countless smart people that will do all the work for us while we will only give them a direction. Just because we don't have them. That means, we are limited in resources (our time) and we should spend it wisely.

Because of all of this I did an experiment, which you can reproduce, and you will supposedly get a similar result.

For the experiment you will need two N900 devices. Boot one to Fremantle and boot the second one to Leste.
Now open two ssh sessions for each device: one is needed for monitoring, and other is for doing actions.

Run the monitoring command on both targets (this one is for measuring battery current):
watch cat /sys/class/power_supply/bq27200-0/current_now

Now you can play with devices to see how the consumed power varies (it's in direct ratio to the current).
Let's do tk (touchscreen + keyboard) lock and see how the current changes. We see that on Fremantle it gradually go to low levels: to about 30 - 70mA. On Leste you can see pretty stable current: 150-180mA.

Now we can try to enable touchscreen and see what changes.
On Fremantle: echo 0 > /sys/devices/platform/omap2_mcspi.1/spi1.0/disable_ts
On Leste: echo 0 > /sys/bus/spi/drivers/tsc2005/spi0.0/disable

In both cases you will see almost no difference in power consumption. If you move you finger on the screen you may notice that the current grows a bit. But you won't see Fremantle to go much above 70 mA. That means it's still in deeper sleep state than Maemo Leste device.

That's all about my experiment.

If you have doubts about the above steps or your practice shows it was erroneous somewhere, please describe the steps needed to reproduce your experiment, so I can reproduce it. Of course your experiment should confirm somehow that your above assumption is correct.

**Comment by MerlijnWajer** [comment]
> We can't put out time and attention everywhere in hope it helps somehow to fix things. It's mandatory to find and fix the real roots of issues. What you are talking about is an assumption that needs to be checked first before we go deeper with it. We need to check somehow that your assumption is correct. Right?

We don't really have to, tmlind also confirmed as much in his droid4-pm script, and I think also saw this happen when I was doing my off-mode tests.

I can share how to do those later, but they require serial. There are also certain registers that can be read to read what components are blocking off-mode (deep sleep mode, basically), but I don't know how to interpret those yet.

**Comment by MerlijnWajer** [comment]
You are probably not seeing the deeper sleeps on Fremantle if you're connected over wifi.

**Comment by spinal84** [comment]
I see ~35mA current. For me it seems to be pretty deep sleep. Considering that WiFi is powered on and I'm connected to the device via and I'm running monitoring command in the real time.
If we calculate 1300/35, we can estimate that device can work about 37 hours in these conditions. 1300 is just an example of the "fully charged" value in mAh for a pretty new battery.
If we take now the lowest value of 150mA of Maemo Leste for the same conditions, we can estimate only ~8 hours of work (1300/150). Which is pretty close to what we see in practice.

**Comment by spinal84** [comment]
> I can share how to do those later, but they require serial.

Isn't it possible to write a script that dumps all needed information to a file (somewhere in /dev/shm) which we can analyze later without a need of having serial?

What I mean is a bash script can power off / power on the screen, it can sleep, it can schedule the WiFi to power on, so we can connect to device and see what's the state of things. It can load and unload the kernel modules.

Is the serial cable really needed for this work?

**Comment by MerlijnWajer** [comment]
I will try to share more info later. But: 

1. Serial is very useful, as it allows booting into a kernel with no modules loaded at, loading the very mininal required, and then seeing the n900 indeed drop into 'off-mode'. I shared this earlier: https://twitter.com/maemoleste/status/1226265997981646849

35mA current is way too much for Fremantle, and it should be able to go way lower.
Both wireless and also USB attached have an effect on power management. Serial does too, but relatively minimally (it will only wake up when you send or receive).

2. The Droid4 might be a better target for you if you want to look at this, since it's relatively easy to make a Droid4 serial, and it's also been tested more thoroughly by Tony.

3. PowerVR on Leste currently blocks any off mode. I've tried to get Leste booted to h-d to enter off-mode, but I had to unload so many modules and it still didn't work. Booting to plan kernel with no modules, I think I did get it to enter low power mode.

I will look more at this in the near future, but have a few other things I want to finish at the moment. When I pick it up next, I will try to be more careful and document what it is that I am testing/doing, so that it is at least reproducible. But I would say, yes, for N900 and Droid 4 power management, you probably want serial. And at least for the Droid 4, that's relatively simple. Let me know if you want me to send you a serial.

**Comment by MerlijnWajer** [comment]
I sent out an email to the ML a few hours ago with a patch, but more work is required: https://lists.dyne.org/lurker/message/20211127.145147.a3d42f2e.en.html

**Comment by MerlijnWajer** [comment]
I also submitted another match that makes the device enter RET/OFF just fine with the module loaded, but there is more to do:

* there seems to be a bug where it tries to reset the ts but can't (likely related to my pm work)
* we might want to not keep the irqs awake all the time if the screen is not used for more than a certain amount of time (autoidle), I think we can do this with `irq_set_irq_wake`
